diff --git a/src/FilenameGenerator.php b/src/FilenameGenerator.php
index 258cf7e..dae7318 100644
--- a/src/FilenameGenerator.php
+++ b/src/FilenameGenerator.php
@@ -53,8 +53,9 @@ class FilenameGenerator implements FilenameGeneratorInterface {
    *   The filename stripped to only safe characters.
    */
   protected function sanitizeFilename($filename, $langcode) {
-    $transformed = $this->transliteration->transliterate($filename, $langcode);
-    return preg_replace("/[^A-Za-z0-9 ]/", '', $transformed);
+    //$transformed = $this->transliteration->transliterate($filename, $langcode);
+    //return preg_replace("/[^A-Za-z0-9 ]/", '', $transformed);
+    return $filename;
   }
 
 }
