diff --git a/src/Plugin/Field/FieldWidget/OptionsShsWidget.php b/src/Plugin/Field/FieldWidget/OptionsShsWidget.php
index 26a9b44..02b492f 100644
--- a/src/Plugin/Field/FieldWidget/OptionsShsWidget.php
+++ b/src/Plugin/Field/FieldWidget/OptionsShsWidget.php
@@ -295,8 +295,7 @@ class OptionsShsWidget extends OptionsSelectWidget implements ContainerFactoryPl
    */
   public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
     if (!isset($values[0]['target_id']) || ($values[0]['target_id'] === '')) {
-      // Return original values.
-      return $values;
+      return null;
     }
 
     $exploded_values = [];
