diff --git a/src/Otp.php b/src/Otp.php
index c87958f..4b52c13 100644
--- a/src/Otp.php
+++ b/src/Otp.php
@@ -150,7 +150,7 @@ class Otp {
       return;
     }
 
-    $accounts = $this->entityTypeManager->getStorage('user')->loadByProperties(['name' => $phone_number]);
+    $accounts = $this->entityTypeManager->getStorage('user')->loadByProperties(['phone_number' => $phone_number]);
     $account = reset($accounts);
     return $account;
   }
@@ -160,7 +160,7 @@ class Otp {
    */
   public function validateOtp($otp, $mobile_number) {
     $users = $this->entityTypeManager->getStorage('user')
-      ->loadByProperties(['name' => $mobile_number]);
+      ->loadByProperties(['phone_number' => $mobile_number]);
     $user = reset($users);
     if ($user) {
       $uid = $user->id();
@@ -225,7 +225,7 @@ class Otp {
    */
   public function validateSessionid($session_id, $mobile_number) {
     $users = $this->entityTypeManager->getStorage('user')
-      ->loadByProperties(['name' => $mobile_number]);
+      ->loadByProperties(['phone_number' => $mobile_number]);
     $user = reset($users);
     if ($user) {
       $uid = $user->id();
@@ -246,7 +246,7 @@ class Otp {
    */
   public function userOtpLogout($session_id, $mobile_number) {
     $users = $this->entityTypeManager->getStorage('user')
-      ->loadByProperties(['name' => $mobile_number]);
+      ->loadByProperties(['phone_number' => $mobile_number]);
     $user = reset($users);
     if ($user) {
       $uid = $user->id();
@@ -383,7 +383,7 @@ class Otp {
    */
   public function userTiniyoOtpLogout($session_id, $mobile_number) {
     $users = $this->entityTypeManager->getStorage('user')
-      ->loadByProperties(['name' => $mobile_number]);
+      ->loadByProperties(['phone_number' => $mobile_number]);
     $user = reset($users);
     if ($user) {
       $uid = $user->id();
